### ISSUE TITLE
Graphql requests are not monitoring friendly

### ISSUE THREAD

**Issue by bcaudan** [apollographql/apollo-client#7130]
**Intended outcome:**

Be able to monitor graphql requests through a third party monitoring library.

**Actual outcome:**

Third party monitoring library override global fetch to gather data on each request.
Since `createHttpLink` keep a reference to fetch ([code](https://github.com/apollographql/apollo-client/blob/d470c964db46728d8a5dfc63990859c550fa1656/src/link/http/createHttpLink.ts#L35)), if global fetch is instrumented after `createHttpLink` execution then graphql requests can't be instrumented.

**How to reproduce the issue:**

with a simple fetch override:
```js
const originalFetch = window.fetch;
window.fetch = function () {
  console.log("request instrumented", arguments[0]);
  return originalFetch.apply(this, arguments);
};
```

if executed **before** `ApolloClient` initialization, graphql requests are logged
if executed **after** `ApolloClient` initialization, graphql requests are not logged

**Versions**

master

**Comment by danmayer** [comment]
This is currently impacting our ability to monitor perf related to Apollo client generated requests. It would be great to come up with a solution that would avoid brittle ordering issues.

**Comment by bignimbus** [comment]
Hi @bcaudan, thanks for your patience!  Would passing the overridden `fetch` function via the `fetch` option in HTTP Link per [these docs](https://www.apollographql.com/docs/react/api/link/apollo-link-http/#fetch) address the issue you're describing?

**Comment by bignimbus** [comment]
Hi @arminrosu 👋🏻 I see you reacted to the above comment - can you elaborate so that we can better understand your use case and narrow down our options here?

**Comment by arminrosu** [comment]
@bignimbus gladly, thanks for the interest.

I'm trying to [set up Datadog tracing](https://docs.datadoghq.com/real_user_monitoring/connect_rum_and_traces/?tab=browserrum) from the client down to our backend services. Datadog does this by setting some headers on GQL requests. They probably do this by wrapping browser requests and appending their headers.

It works if I initialise datadog _before_ the apollo client. But then I started having issues with Apollo Cache. I dropped the topic for now, but that's essentially it. 

I haven't tried setting a custom `fetch` instance, as I [don't see it exported](https://github.com/DataDog/browser-sdk/tree/3569ad9435de867c5b9ecff10f1c2ce136e4ea85/packages/rum-core/src).

**Comment by bignimbus** [comment]
Thanks @arminrosu - looking at the Datadog SDK now.

https://github.com/DataDog/browser-sdk/blob/46c51d01d497128d61532043048fde15d535c331/packages/core/src/browser/fetchObservable.ts#L41-L70

This snippet was interesting to me.  It does seem like something that would be resolved by, as you said, initializing the Datadog SDK before Apollo Client.

> But then I started having issues with Apollo Cache

That's also interesting and something that might be worth getting more detail on - what issues did you see?  Is this a bug in Apollo Client?

**Comment by arminrosu** [comment]
> That's also interesting and something that might be worth getting more detail on - what issues did you see? Is this a bug in Apollo Client?

I'll get back to you once I give this one another shot. I haven't had time to look into it closely yet.

**Comment by alessbell** [comment]
Hi @arminrosu 👋 Just following up here and was curious if there are any more details you or other participants on this issue can provide to @bignimbus's question above re: downsides to initializing DD SDK before Apollo Client. Thanks!

**Comment by arminrosu** [comment]
> downsides to initializing DD SDK before Apollo Client

@alessbell thanks for checking in. No complaints from doing this. We now have it in production for a few months and everything is working fine.

**Comment by phryneas** [comment]
I believe this issue should have been solved with #8603 and only occur in Apollo Client versions before [v3.4.6](https://github.com/apollographql/apollo-client/releases/tag/v3.4.6).

As a result, I'm going to close this issue. 
If it is still occurring in newer versions, please give this a bump so I'll reopen, or open a new issue :)
